app_dart/lib/src/request_handlers/get_benchmarks.dart

```diff
@@ -30,26 +33,102 @@ class GetBenchmarks extends RequestHandler<Body> {
+          ? maxRecords - releaseBranchCommits.length
+          : 0;
+      timestamp = releaseBranchCommits.last.timestamp;
+    }
+
+    /// Query all remaining commits from master.
+    if (branch == master || masterLimit > 0) {
+      /// `+1` is to guarantee picking up the master commit, from which
+      /// the release branch is derived.
+      final List<Commit> masterCommits = await datastore
+          .queryRecentCommits(timestamp: timestamp + 1, limit: masterLimit)
+          .toList();
+
+      masterMap = await _getBenchmarks(
+          masterLimit, master, datastore, masterCommits, benchmarks, timestamp);
+    }
+
+    _combineValues(releaseBranchMap, masterMap, benchmarks);
+
+    return Body.forJson(<String, dynamic>{
+      'Benchmarks': benchmarks,
+    });
+  }
+
+  /// Combine results for both release and master branches.
+  void _combineValues(Map<String, Result> releaseBranchMap,
+      Map<String, Result> masterMap, List<Map<String, dynamic>> benchmarks) {
     final KeyHelper keyHelper = config.keyHelper;
+    final Map<String, Result> map =
+        releaseBranchMap.isNotEmpty ? releaseBranchMap : masterMap;
+    for (String task in map.keys) {
+      final List<TimeSeriesValue> timeSeriesValues = <TimeSeriesValue>[];
 
-    final List<Map<String, dynamic>> benchmarks = <Map<String, dynamic>>[];
-    final Set<Commit> commits = await datastore
-        .queryRecentCommits(limit: maxRecords, branch: branch)
-        .toSet();
-    await for (TimeSeries series in db.query<TimeSeries>().run()) {
-      final Query<TimeSeriesValue> query =
-          db.query<TimeSeriesValue>(ancestorKey: series.key)
-            ..filter('branch =', branch)
-            ..order('-createTimestamp')
-            ..limit(maxRecords);
+      timeSeriesValues.addAll(releaseBranchMap.containsKey(task)
+          ? releaseBranchMap[task].timeSeriesValues ?? <TimeSeriesValue>[]
+          : <TimeSeriesValue>[]);
+      timeSeriesValues.addAll(masterMap.containsKey(task)
+          ? masterMap[task].timeSeriesValues ?? <TimeSeriesValue>[]
+          : <TimeSeriesValue>[]);
+
+      benchmarks.add(<String, dynamic>{
+        'Timeseries': <String, dynamic>{
+          'Timeseries': map[task].timeSeries,
+          'Key': keyHelper.encode(map[task].timeSeries.key)
+        },
+        'Values': timeSeriesValues,
+      });
+    }
+  }
 
+  Future<Map<String, Result>> _getBenchmarks(
+      int limit,
+      String branch,
+      DatastoreService datastore,
+      List<Commit> commits,
+      List<Map<String, dynamic>> benchmarks,
+      int timestamp) async {
+    final Map<String, Result> map = <String, Result>{};
+
+    await for (TimeSeries series in datastore.db.query<TimeSeries>().run()) {
       final Map<String, TimeSeriesValue> valuesByCommit =
           <String, TimeSeriesValue>{};
-      await for (TimeSeriesValue value in query.run()) {
+
+      /// Adding `1` hour to [timestamp] to guarantee all [values] belonging to
```

Comment:
1 hour is too much, how did we get to this value?

Reply:
Like already mentioned in the comment doc: `values` are inserted into datastore later than `commit`. The workflow inserts `commit` first into datastore, suppose with `timestamp1`, and then inserts `values` of different tests results later once tests finish, suppose with `timestampX`. `timestampX` varies for different tests/commits, depending on test execution time and environment. Based on our infra metrics collected/analyzed, executing time of a commit row is less than 1 hour, that's why I put the 1 hour timeframe.  If too small, we may miss some `values`.

Reply:
Adding, subtracting time can cause race conditions(we don't control the time) should we fix the query to return only commits that have all the data instead?

Reply:
I guess the current logic is designed like this to reduce datastore traffic. If we query timeseriesValue based on commit, then we need to make `config.maxRecords` requests each time the web refreshes.

With current logic, it needs to query datastore only once. 

Also, I don't think race condition is an issue here. This API reads data all the time, rather than writes anything. If data is not there yet, it simply show empty value on the web.

Reply:
My suggestion was to update the query to return only commits with complete results up to the passed timestamp. That should not affect the current performance and will eliminate the need to guess if a commit is complete based on estimates of the execution time.

Reply:
Unluckily that is not possible due to datastore query limitation.
To achieve what you suggested, we need to add a `commit` list which contains all commits to filter out our interested records of timeseriesValue. However, dart db supports only simple filter options like `=` (https://github.com/dart-lang/gcloud/blob/master/lib/src/db/db.dart#L128).

Reply:
Added a TODO: https://github.com/flutter/flutter/issues/56731